Modules/_datetimemodule.c

```diff
@@ -1521,20 +1521,15 @@ wrap_strftime(PyObject *object, PyObject *format, PyObject *timetuple,
     if (newfmt == NULL) goto Done;
     pnew = PyBytes_AsString(newfmt);
     usednew = 0;
+    ch = *pin;
 
-    while ((ch = *pin++) != '\0') {
-        if (ch != '%') {
+    do {
```

Comment:
Hm... It's a bit hard to keep track of where null checking happens in this loop.

I guess the reason that trailing `%` was forbidden in the original version was that they wanted to check for null termination every time the `pin` pointer is advanced.

I think there's a change in behavior around null-checking here, because before this change, the new format string did not include the null terminator, after the change, the new format string does include the null terminator. I don't know that this is a *problem*, but I'm wary of it.

Maybe you can change it like this:

```c
while((ch = *pin++) != '\0') {
    if (ch != '%' || (ch = *pin++) == '\0') {
        ptoappend = pin - 1;
        ntoappend = 1;
    }
    else if (ch == 'z') {
    }
```

That should keep

Reply:
As an aside, this terse C-style stuff where you have to keep track of null pointers is for the birds. It took me a long time to convince myself that there was a change in behavior, and even still I am not certain about it. It really feels weird to advance the pointer *in the conditionals*.

Reply:
Just looking at it, I don't _think_ your code does what you think it does. Without a null char check at the end of the loop, you'll read the null char and advance the pointer, and then the check in the while loop condition will actually be reading the char _after_ the null char.

Like you said though, this stuff is tricky, so I'm going to test it and report back.

Also, my first attempt at fixing this, while I think a bit uglier than the do/while solution, comes closer to your desired behavior then the current one. Here it is:

https://github.com/python/cpython/blob/d178065d5a4e76c00bb1a14dd3d20e935f913537/Modules/_datetimemodule.c#L1525-L1626

As far as the pointer juggling in the conditionals goes, it's a pattern I've seen before in C parsing code. I think ideally you only mutate the pointer in one place, as it quickly becomes hard to reason about.

Reply:
As I suspected, your code doesn't do what we want. Passing it "%D %" produces the format string `b'%D \x00\xfb\xfb\xfb\xfb\xfb\xfb\xfb\xfb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb'`.

I obtained this by calling `printf("%s\n", PyUnicode_AsUTF8(PyObject_Repr(newfmt)));` after the while loop. If you put the call in the loop itself, you can watch it happen: 
```
b'%D\xcb\xcb\xcb'
b'%D \xcb\xcb'
b'%D \x00\xcb'
b'%D \x00\xfb'
b'%D \x00\xfb\xfb\xcb\xcb\xcb\xcb'
b'%D \x00\xfb\xfb\xfb\xcb\xcb\xcb'
b'%D \x00\xfb\xfb\xfb\xfb\xcb\xcb'
b'%D \x00\xfb\xfb\xfb\xfb\xfb\xcb'
b'%D \x00\xfb\xfb\xfb\xfb\xfb\xfb'
b'%D \x00\xfb\xfb\xfb\xfb\xfb\xfb\xfb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb'
b'%D \x00\xfb\xfb\xfb\xfb\xfb\xfb\xfb\xfb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb'
```

I'm not sure what's causing it to eventually break out of the loop and not spin forever; I assume it's one of the length checks that happens in the loop.

Reply:
@MichaelSaah Good call, sorry, I didn't actually try my version, you are right that I advance but then never check if the next character is a `NULL`. I think maybe this will do what I was looking for?

```c
while((ch = *pin++) != '\0') {
    if (ch != '%' || *pin == '\0') {
        ptoappend = pin - 1;
        ntoappend = 1;
    }
    else if (ch == 'z') {
    }
```